### PR TITLE
Grist Toasts

### PR DESCRIPTION
**Pull request by seriousCoolness**
Converted the old text-based grist notifications into Toasts, like how advancements or recipe notifs are handled.

### CODE REVIEW

**Comment by seriousCoolness** [comment]
I don't know why it wants to merge those commits by dweblenod. I could have sworn i based this branch off of the current main 1.18 branch.

**Comment by seriousCoolness** [comment]
Oh, also important to note that they're still only done client-side. So no information is sent to the server, meaning that recipients of /sendgrist won't get toast-notifications, and Sburb client-players still won't get notified that their server-players are using their grist.

**Comment by seriousCoolness** [comment]
> Oh, also important to note that they're still only done client-side. So no information is sent to the server, meaning that recipients of /sendgrist won't get toast-notifications, and Sburb client-players still won't get notified that their server-players are using their grist.

Nevermind. Making them networked was way easier than I thought it would be. They work across clients now.

**Comment by kirderf1** [comment]
> I don't know why it wants to merge those commits by dweblenod. I could have sworn i based this branch off of the current main 1.18 branch.

Your branch appears to be built on dweblenods 1.18 branch, which contains commits that were squashed when merged into mraof:1.18. Dweblenod shouldn't have kept those commits alive through that branch after it was squash-merged, but it is what it is.
If you're smart with git, you could drop the commits in question and force push your own commits without dweblenods. Otherwise, you can leave it as is. The changes from those commits no longer show up under "Files Changed" after you merged mraof:1.18.

**Comment by kirderf1** [comment]
You appear to be using a string to represent the notification source. However as there are only a fixed few possible sources (Client, Server and SendGrist), I think it makes more sense to represent it with an enum.

**Comment by seriousCoolness** [comment]
> You appear to be using a string to represent the notification source. However as there are only a fixed few possible sources (Client, Server and SendGrist), I think it makes more sense to represent it with an enum.

Done. Hopefully I did this correctly.

File: src/main/java/com/mraof/minestuck/network/GristToastPacket.java
Comment (Dweblenod) [inline src/main/java/com/mraof/minestuck/network/GristToastPacket.java:45]:
```
Player playerEntity = ClientProxy.getClientPlayer();
		
if(playerEntity != null)
{
	GristSet gristValue = this.gristValue;
	GristToast.EnumSource source = this.source;
	boolean increase = this.increase;
	GristHelper.sendGristMessage(playerEntity, gristValue, source, increase);
}
```

**Review by Dweblenod — COMMENTED** [review]
Its already been noted outside of Git, but a crash will occur on loaded dedicated servers if you try to get the player directly from Minecraft. Gave a recommendation to fix it.

Other than that, I don't see any issues with the code and I have had no issues with it in game(aside from the crash). There's are few sections of code that use unnecessary calls or could otherwise be simplified, but I will defer to Kirderf for the specifics of formattings

File: src/main/java/com/mraof/minestuck/network/GristToastPacket.java
Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/network/GristToastPacket.java]: Fixed. Thanks again, Dweble.

File: src/main/java/com/mraof/minestuck/alchemy/GristHelper.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/alchemy/GristHelper.java:163]: This should always be false when you're getting the player from a server via a player identifier.

File: src/main/java/com/mraof/minestuck/alchemy/GristHelper.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/alchemy/GristHelper.java:158]: The edit data was previously used to send a grist notification to the edit player. It looks like you're now instead only sending grist notifications to the player which the grist cache belongs to.

File: src/main/java/com/mraof/minestuck/alchemy/GristHelper.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/alchemy/GristHelper.java:170]: `playerIdentifier.getPlayer()` might return null. You need to account for this.

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java:17]: These annotations are not meant for mod code. Remove this.

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java:36]: This enum is not used client-side only, so I suggest moving it elsewhere, such as to `GristHelper`

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java:39]: Some braces are not on new lines when they should be. Make sure to reformat this class so that it matches the code style that is set for this project.

File: src/main/java/com/mraof/minestuck/alchemy/GristHelper.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/alchemy/GristHelper.java:144]: Basic javadoc on this function seems appropriate.

File: src/main/java/com/mraof/minestuck/alchemy/GristHelper.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/alchemy/GristHelper.java:173]: Commented old code that should be removed.

File: src/main/java/com/mraof/minestuck/alchemy/GristHelper.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/alchemy/GristHelper.java:178]: This function is more appropriate for `GristToast`.

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java:18]: A short javadoc seems appropriate on this class.

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java:35]: A simple comment or javadoc seems appropriate for each of these values.

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java:83]:
```suggestion
		if(this.increase)
```

**Review by kirderf1 — CHANGES_REQUESTED** [review]
I've been pretty thorough, but not completely thorough. Primarily, I haven't reviewed `GristToast` properly, but am thinking that I'll do that in the next review.

File: src/main/java/com/mraof/minestuck/command/SendGristCommand.java
Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/command/SendGristCommand.java:47]: There are certain cases where you shouldn't send notifications, or notifications would be redundant. (Like, when using the /grist command, it already notifies the player through the chat.) I think it's better to leave them separate, at least for now.

File: src/main/java/com/mraof/minestuck/alchemy/GristHelper.java
Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/alchemy/GristHelper.java]: Ah, okay. It's been removed.

File: src/main/java/com/mraof/minestuck/alchemy/GristHelper.java
Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/alchemy/GristHelper.java]: Okay, cool. I thought that was why it was there, but I wasn't sure. It's been removed.

File: src/main/java/com/mraof/minestuck/alchemy/GristHelper.java
Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/alchemy/GristHelper.java:158]: I put it in GristHelper because GristSet.getMap() is protected for some reason, and I think it's easier to read this way.

File: src/main/java/com/mraof/minestuck/command/SendGristCommand.java
Comment (kirderf1) [reply src/main/java/com/mraof/minestuck/command/SendGristCommand.java:47]: `/grist` doesn't notify the receivers, only the command user. `/sendgrist` however notifies both the receiver and sender.

Comment (kirderf1) [reply src/main/java/com/mraof/minestuck/command/SendGristCommand.java:47]: Yeah, I suppose it makes sense to not modify `increase()` or `decrease()`. However, doing so but with two new functions would be nice. For example, such that the code that this comment is attached to could be replaced with:
```java
GristHelper.decreaseAndNotify(player.level, IdentifierHandler.encode(player), grist, GristToast.EnumSource.SENDGRIST);
```

File: src/main/java/com/mraof/minestuck/alchemy/GristHelper.java
Comment (kirderf1) [reply src/main/java/com/mraof/minestuck/alchemy/GristHelper.java:158]: You could use `gristSet.getAmounts()` instead of the map.

Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/alchemy/GristHelper.java]: Alright, it's been moved to GristToast.

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java:176]: You're not actually using the player argument for anything but the `isLocalPlayer()` check. You could probably remove the check entirely, but if you still want a check to that effect, you might be able to use `minecraft.isSameThread()`?

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java:188]: Replace 01 with either 1 or 0.

**Review by kirderf1 — CHANGES_REQUESTED** [review]
There are some comments from earlier that you might've missed. Make sure to look at those as well!

One motivation for the original version of grist notifications was to notify players in editmode so that they're made aware when one of their actions spend grist. However grist notifications are no longer sent to the editmode player, assuming they're not also the player that the grist cache belongs to. Will you add back something such that an editmode player gets notified of grist changes as well?

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java]: Should be gone in the latest commits.

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java]: I think it should be good now.

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java]: It was mostly just a sanity check, so I've removed it.

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java]: It was a lowercase L not a 1, but either way it is unnecessary. Fixed.

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java]: I decided to go with the switch statement. Done.

File: src/main/java/com/mraof/minestuck/alchemy/GristHelper.java
Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/alchemy/GristHelper.java]: Done. Feel free to tell me if you think it could use clarification.

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java:26]:
```suggestion
	private static final ResourceLocation TEXTURE = new ResourceLocation("minestuck", "textures/gui/toasts.png");
```

File: src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/client/gui/toasts/GristToast.java:91]:
```suggestion
		switch(this.source)
		{
			case CLIENT -> pToastComponent.blit(pPoseStack, 5, 5, 196, 0, 20, 20);
			case SERVER -> pToastComponent.blit(pPoseStack, 5, 5, 196, 20, 20, 20);
			case SENDGRIST -> pToastComponent.blit(pPoseStack, 5, 5, 216, 0, 20, 20);
			case CONSOLE -> pToastComponent.blit(pPoseStack, 5, 5, 216, 20, 20, 20);
		}
```

File: src/main/java/com/mraof/minestuck/computer/editmode/ServerEditHandler.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/computer/editmode/ServerEditHandler.java:435]: Three comments that applies to all of these new notifications:
- You can compare player identifiers directly with `==` and `!=`.
- You're notifying the server player. However who you want to notify is the edit player, which is usually, but *not necessarily* the same player as the server player.
- This seems to apply to all notifications with the `SERVER` source. Do you think it would make sense to implement this extra notification into `GristHelper.notify()` or such instead?

File: src/main/java/com/mraof/minestuck/network/GristToastPacket.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/network/GristToastPacket.java:43]: Player entity is not being used in any meaningful way. You can remove it.

**Review by kirderf1 — COMMENTED** [review]
Not much left now. Just one significant comment on the new notifications along with some minor code cleanup comments.

File: src/main/java/com/mraof/minestuck/computer/editmode/ServerEditHandler.java
Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/computer/editmode/ServerEditHandler.java]: Moved it all into notify() and changed the player from the server to the editmode user.

File: src/main/java/com/mraof/minestuck/network/GristToastPacket.java
Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/network/GristToastPacket.java]: Got it.

File: src/main/java/com/mraof/minestuck/item/block/MultiblockItem.java
Comment (kirderf1) [inline src/main/java/com/mraof/minestuck/item/block/MultiblockItem.java:86]: You used an inappropriate method when merging mraof:1.18, which copied bae517355f43b4fd978551119bfac19746e9bdbf and 9c4d3b61c16481ad70c8ec4a85847309553f9722 as new commits ba19ed26f199ba00d28ce5727def2163cba5dba5 and 2feb29898ed51e677fe905ab50a3c57b78da1b7b. While nothing need to be done about it this time, make sure next time to perform a regular merge (or possibly a rebase as long as you understand the consequences of it).

Comment (seriousCoolness) [reply src/main/java/com/mraof/minestuck/item/block/MultiblockItem.java:86]: I thought I did a regular merge, but I might have made a mistake when using intelliJ's built-in git features.

**Review by kirderf1 — APPROVED** [review]

**Pull request merged**
